Auth features for server (#9190)
bokeh/bokeh: sphinx/source/docs/user_guide/server.rst

Purpose
-------

> The purpose of the Bokeh server is to make it easy for python users to create

Reviewer: ```suggestion
The purpose of the Bokeh server is to make it easy for Python users to create
```

    }

.. _userguide_server_deployment_auth:

Authentication
~~~~~~~~~~~~~~

The Bokeh server itself does not have any facilities for authentication or

> authorization. However, the Bokeh server can be configured with and "Auth

Reviewer: ```suggestion
authorization. However, the Bokeh server can be configured with an "Auth
```

Provider" that hooks in to Tornado's underlying capabilities. For background
information, see the Tornado docs for `Authentication and security`_. The rest

> of this section assume some familiarity with that material.

Reviewer 2: ```suggestion
of this section assumes some familiarity with that material.
```

Auth Module
'''''''''''

The Bokeh server can be configured to only allow connections in case there is

> a properly authenticed user. This is accomplished by providing the path to

Reviewer: ```suggestion
a properly authenticated user. This is accomplished by providing the path to
```

a module that implements the necessary functions on the command line:

.. code-block:: sh

    bokeh serve --auth-module=/path/to/auth.py

or by setting the ``BOKEH_AUTH_MODULE`` environment variable.

The module must contain *one* of the following two functions that will return
the current user (or None):

.. code-block:: python

    def get_user(request_handler):
        pass

>     async def get_user_aync(request_handler):

Reviewer 2: ```suggestion
    async def get_user_async(request_handler):
```

        pass

The function is passed the Tornado ``RequestHandler`` and can inspect cookies
or request headers to determine the authenticated user. If there is no valid
authenticated user, these functions should return None.

Additionally, the module must specify where to redirect unauthenticated users.
It must contain either:

* a module attribute ``login_url`` and (optionally) a ``LoginHandler`` class
* a function definition for ``get_login_url``

.. code-block:: python

    login_url = "..."

    class LoginHandler(RequestHandler):
        pass

    def get_login_url(request_handler):
        pass

When a relative ``login_url`` is given, an optional ``LoginHandler`` class may
also be provided, and it will be installed as a route on the Bokeh server
automatically.

The ``get_login_url`` function is useful in cases where the login URL must
vary based on the request, or cookies, etc. It is not possible to specify a
``LoginHandler`` when ``get_url_function`` is defined.

Analogous to the login options, optional ``logout_url`` and ``LogoutHandler``

> values may be define and endopoint for logging users out.

Reviewer 2: ```suggestion
values may be define an endpoint for logging users out.
```

    run the Bokeh server behind a proxy that terminates SSL, so that the session
    ID is transmitted securely to the user's browser.

XSRF Cookies
''''''''''''

Bokeh can enable the use of Tornado's cross-site request forgery protection
protection. To turn this feature on, use the ``--enable-xsrf-cookies`` option,
or set the environment variable ``BOKEH_XSRF_COOKIES=yes``. If this setting is
enabled, any PUT, POST, or DELETE operations on custom or login handlers must be
instrumented properly in order to function. Typically, this means adding the
code:

.. code-block:: html

    {% module xsrf_form_html() %}

> to all HTML form submissions templates. For full details, see the Tornado

Reviewer: ```suggestion
to all HTML form submission templates. For full details, see the Tornado
```
